File: src/dotnet/Orders.Api/Jobs/CheckShippingStatus.cs
using Orders.Application.Interfaces;
using Orders.Application.Specifications.Orders;
using Orders.Domain.Models;

namespace Orders.Api.Jobs;

public class CheckShippingStatus : IShippingStatusJob
{
    private readonly ICarrierService _carrierService;
    private readonly ILogger<CheckShippingStatus> _logger;
    private readonly IRepository<Order> _orderRepository;
    private readonly IRepository<ShippingDetails> _shippingDetailsRepository;

    public CheckShippingStatus(ILogger<CheckShippingStatus> logger, IRepository<Order> orderRepository,
        IRepository<ShippingDetails> shippingDetailsRepository, ICarrierService carrierService)
    {
        _logger = logger;
        _orderRepository = orderRepository;
        _shippingDetailsRepository = shippingDetailsRepository;
        _carrierService = carrierService;
    }

    public async Task CheckStatus()
    {
        int ordersShipped = 0;
        int ordersDelivered = 0;

        var notShippedOrders = await _orderRepository.ListWithSpecification(new OrdersThatAreNotShippedSpec());
        var notDeliveredOrders = await _orderRepository.ListWithSpecification(new OrdersThatAreNotDeliveredSpec());

        foreach (var order in notShippedOrders)
        {
            if (!_carrierService.IsOrderShipped(order.Id))
                continue;

            order.ShippingDetails!.ShippedAt = DateTime.UtcNow;
            await _shippingDetailsRepository.Update(order.ShippingDetails);
            ordersShipped++;
        }

        foreach (var order in notDeliveredOrders)
        {
            if (!_carrierService.IsOrderDelivered(order.Id))
                continue;

            order.ShippingDetails!.DeliveredAt = DateTime.UtcNow;
            await _shippingDetailsRepository.Update(order.ShippingDetails);
            ordersDelivered++;
        }

        _logger.LogInformation("Amount of orders shipped: {OrdersShipped} out of {TotalOrders}", ordersShipped,
            notShippedOrders.Count);
        _logger.LogInformation("Amount of orders delivered: {OrdersDelivered} out of {TotalOrders}", ordersDelivered,
            notDeliveredOrders.Count);
    }
}

File: src/laravel/app/Jobs/CheckShippingStatus.php
<?php

namespace App\Jobs;

use App\Models\Order;
use App\Services\Interfaces\ICarrierService;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldBeUnique;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;

class CheckShippingStatus implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    private ICarrierService $carrierService;

    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct()
    {
        $this->carrierService = app()->make(ICarrierService::class);
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle(): void
    {
        $ordersShipped = 0;
        $ordersDelivered = 0;

        $notShippedOrders = Order::with('shippingDetails')
            ->whereRelation('shippingDetails', 'shipped_at', '=', null)
            ->get();

        $notDeliveredOrders = Order::with('shippingDetails')
            ->whereRelation('shippingDetails', 'shipped_at', '!=', null)
            ->whereRelation('shippingDetails', 'shipped_at', '<', now()->addDay())
            ->whereRelation('shippingDetails', 'delivered_at', '=', null)
            ->get();

        $notShippedOrders->each(function (Order $order) use (&$ordersShipped) {
            if (!$this->carrierService->isOrderShipped($order)) {
                return;
            }

            $order->shippingDetails->shipped_at = now();
            $order->shippingDetails->save();
            $ordersShipped++;
        });

        $notDeliveredOrders->each(function (Order $order) use (&$ordersDelivered) {
            if (!$this->carrierService->isOrderDelivered($order)) {
                return;
            }

            $order->shippingDetails->delivered_at = now();
            $order->shippingDetails->save();
            $ordersDelivered++;
        });

        \Log::info("Orders shipped: {$ordersShipped}");
        \Log::info("Orders delivered: {$ordersDelivered}");
    }
}
